### DISCUSSION TITLE
Developing a more sustainable method of interacting with Jupyter notebooks when posting articles

### DISCUSSION THREAD

**Discussion by anth-volk** [PolicyEngine/policyengine-app#1466] [Ideas]
In getting the opportunity to work with Nikhil this week on a couple articles, I think our system of using Jupyter notebooks is an ingenious way of harnessing the power of Python without needing complicated embeds or the use of third-party tools like Streamlit. However, a couple downsides jumped out at me, and I think we could handle these negatives in order to strengthen how we utilize Jupyter notebooks moving forward.

* Notebook execution is completely dependent on one's local environment, which created issues when I had packages that weren't fully up-to-date, but wasn't aware of it because there's nothing akin to the `package-lock.json` file present in NPM development environments. It also meant that I faced strange issues creating the table in the second article.
* Charts created with the notebooks are a bit different visually from those we create within the app, creating inconsistent branding. This could be relatively easy to solve if we created a series of standard styling settings. Additionally, the charts lack the PolicyEngine watermark that the app adds, which I think would be beneficial to have.
* Creating charts requires a series of non-obvious imports from different packages. While Nikhil's efforts to create a standardized package from which to import seems to have helped quite a bit, I think we can go a bit further still.
* Certain charts are utilized in just about every article we write, but we still have to manually create these. I could imagine creating a utility function, for example, that's passed args representing country ID, married/single, and number of children, a reform, and maybe something else, and it outputs a "varying your earnings" chart in a standardized color scheme. There could even be an escape hatch to allow manual manipulation of plotly args or something like that.

**Comment by MaxGhenis** [comment]
Not sure if these address the issues you raise, but various packages out there aid in publishing notebooks online, e.g. [quarto](https://nbdev.fast.ai/tutorials/blogging.html)

**Comment by daemon-reconfig** [comment]
Can we integrate google colab?

**Comment by anth-volk** [comment]
@daemon-reconfig I believe this is actually something we tried a while back, but couldn't get working, but others can correct me if I'm wrong. The problem is that Google just doesn't want other apps linking with Colab, or so it would seem.

**Comment by PradyXd** [comment]
how about quilljs?

**Reply by anth-volk** [reply]
Interesting, will have to check out! Thanks for raising it as an idea.
